### PR TITLE
Encryption: Make DEKs cache TTL & cleanup interval configurable

### PR DESCRIPTION
**Pull request by joanlopez**
**What this PR does / why we need it**:

Make the default TTL value for new entries into the data encryption keys cache a setting configurable through Grafana configuration. Currently it's a constant value and we need to be able to configure it for certain use cases.

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana-enterprise/issues/2916

**Special notes for your reviewer**:

I experienced some `concurrent map writes` while doing some tests so I decided to add a `sync.RWMutex` to prevent that kind of issues, then decided to abstract cache details into a specific struct to keep the code a bit easier to read.

Additional comments inline.

### CODE REVIEW

File: pkg/services/secrets/manager/manager_test.go [hunk 1]
```diff
@@ -300,23 +294,7 @@ func TestSecretsService_Run(t *testing.T) {
 		// Then, once the ticker has been triggered,
 		// the cleanup process should have happened,
 		// therefore the cache should be empty.
-		require.Len(t, svc.dataKeyCache, 0)
-	})
-
-	t.Run("should update data key expiry after every use", func(t *testing.T) {
```
Comment (joanlopez) [inline pkg/services/secrets/manager/manager_test.go:306]: I intentionally removed that behavior cause we've explicitly been requested to add support for the opposite behavior (not TTL renewals) and I see no strong reasons to keep support for both strategies (code would become more complex, etc with no clear benefits).

In terms of performance, I think the overhead of an extra network request (to decrypt the data encryption key against the external key vault) every fifteen minutes (or so) should be more than acceptable.

File: pkg/services/secrets/manager/manager_test.go [hunk 2]
```diff
@@ -280,18 +280,12 @@ func TestSecretsService_Run(t *testing.T) {
...
-		// the ticker is triggered.
-		gcInterval = time.Nanosecond
+		require.Len(t, svc.dataKeyCache.entries, 1)
 
 		t.Cleanup(func() { now = time.Now })
-		now = func() time.Time { return time.Now().Add(dekTTL) }
+		now = func() time.Time { return time.Now().Add(cacheTTLByDefault) }
 
-		ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+		ctx, cancel := context.WithTimeout(ctx, 1*time.Second)
```
Comment (joanlopez) [inline pkg/services/secrets/manager/manager_test.go:288]: Apparently the previous value (100ms) made that test a bit flaky.
So, I think x10ing the value should be enough (🤞🏻 ) and not a big deal.

I'm also open to other suggestions, but I cannot see an easy way to test that feature without big code changes just for the sake of testing.

File: conf/defaults.ini [hunk 3]
```diff
@@ -295,6 +295,14 @@ content_security_policy_template = """script-src 'self' 'unsafe-eval' 'unsafe-in
 # Controls if old angular plugins are supported or not. This will be disabled by default in Grafana v9.
 angular_support_enabled = true
 
+[security.encryption]
```
Comment (joanlopez) [inline conf/defaults.ini:298]: I think we're already using this section for Enterprise features, but it shouldn't cause any conflict, right? 🤔

Comment (undef1nd) [reply conf/defaults.ini:298]: I think it's fine.

**Review by undef1nd — APPROVED** [review]
Looks good!

**Review by xlson — APPROVED** [review]
LGTM!

File: conf/defaults.ini (hunk 3 above)
Comment (joanlopez) [reply conf/defaults.ini:298]: Yeah, I did some tests and apparently the `gopkg.in/ini.v1` package is able to automatically merge fields from same section. 

So, something like:
```
[security.encryption]
data_keys_cache_ttl = 15m

data_keys_cache_cleanup_interval = 1m

# Enterprise only
[security.encryption]
algorithm = "rsa-sha256"
```

Ends up with the `security.encryption` initialized with these three values.

Is that what you meant? @xlson  🤔  Thanks!!

Comment (xlson) [reply conf/defaults.ini:298]: That's what I meant, great! I'll mark this conversation as resolved then.

File: pkg/services/secrets/manager/helpers.go [hunk 4]
```diff
@@ -22,7 +22,11 @@ func SetupTestService(tb testing.TB, store secrets.Store) *SecretsService {
 	}
 	raw, err := ini.Load([]byte(`
 		[security]
-		secret_key = ` + defaultKey))
+		secret_key = ` + defaultKey + `
+
+		[security.encryption]
+		data_keys_cache_ttl = 5m
+		data_keys_cache_cleanup_interval = 1ns`))
```
[the remark this answers is not in the document]
Comment (xlson) [reply pkg/services/secrets/manager/helpers.go:29]: no that's fine, just wanted to make sure that it wasn't a mistake :)

**Pull request merged**